### ISSUE TITLE
[LICM] LICM miscompiles IR (introduces UB) when constant folding (regression since LLVM 17)

### ISSUE THREAD

**Issue by danilaml** [llvm/llvm-project#85448]
LICM transforms the following IR:
```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128-ni:1-p2:32:8:8:32-ni:2"
target triple = "x86_64-unknown-linux-gnu"

define void @test(i32 %a) {
entry:
  br label %loop

loop:                              ; preds = %loop, %entry
  %val = phi i32 [ 1, %loop ], [ 0, %entry ]
  %0 = or i32 %val, %a
  %1 = mul nuw i32 %0, 65536
  %sext = mul nsw i32 %1, -9
  %2 = call i32 null(i32 %sext)
  br label %loop
}
```
into
```llvm
define void @test(i32 %a) {
entry:
  br label %loop

loop:                                             ; preds = %loop, %entry
  %val = phi i32 [ 1, %loop ], [ 0, %entry ]
  %0 = or i32 %val, %a
  %1 = mul nuw i32 %0, -589824
  %2 = call i32 null(i32 %1)
  br label %loop
}
```

Which is not equivalent. It needs to drop `nuw` flag on mul, since now %1 might be overflowing (suppose it was value close to 2^16 - 1 before).

Godbolt: https://godbolt.org/z/Wh9G8nsrr
Alive2: https://alive2.llvm.org/ce/z/admTj4

**Comment by danilaml** [comment]
Bisect pointed to https://github.com/llvm/llvm-project/commit/2dd52046816ac706a25d588efac77705793d8778

**Comment by topperc** [comment]
The title says LLVM 17, but I guess it should be LLVM 18 since `https://github.com/llvm/llvm-project/commit/2dd52046816ac706a25d588efac77705793d8778` isn't in 18?

**Comment by topperc** [comment]
> @topperc I meant that regression in behavior compared to LLVM 17 (since I haven't done the bisect yet at that point and just confirmed that LLVM 17.0.1 is not affected). I see the ambiguity.

FYI I filed an issue with compiler explorer about 18.1.0 tools being missing for `LLVM IR` sources

**Comment by topperc** [comment]
Should be fixed after [1261c02](https://github.com/llvm/llvm-project/commit/1261c02be4829984f025f258f3717277d04a8332)
